### PR TITLE
Added plugin upload functionality

### PR DESCRIPTION
**Pull request by zhongruige**
### Changes proposed in this Pull Request:

The goal of this PR is to allow installing the WooCommerce plugin. It makes the following changes:

* adds `UPDATE_WC` as an environment variable to determine if the update WooCommerce test should run
* updates to `smoke-test-daily.yml` to run the plugin upload utility for WooCommerce and sets the `UPDATE_WC` environment variable to true
* `get-plugin-zip.js` in the environment package for handling getting the zip and checking if it has a nested zip.
* three utility methods: `getSlug()` to get the slug string for a plugin; `describeIf()` to conditionally run a test suite; `itIf()` to conditionally run a test case.
* merchant workflows around plugin work: `uploadAndActivatePlugin()`, `activatePlugin()`, `deactivatePlugin()`, `deletePlugin()`

### How to test the changes in this Pull Request:

1. Review changes and the updates to the README and change log files
2. Run `npm install`
2. Run `npx wc-e2e docker:up`
3. Run `npx wc-e2e test:e2e`
4. Verify tests pass

##### Test plugin workflow

* Update `update-woocommerce.test.js` with new values for:
  * `nightlyZip`
  * `pluginName`
* For example, in my testing I used:
 
```
const nightlyZip = 'https://github.com/Automattic/woocommerce-payments/releases/download/2.6.1/woocommerce-payments.zip';
const pluginName = 'WooCommerce Payments';
```

⚠️ Make sure the above is done first before running the command below, otherwise this will break the local test environment.

After the above is done, run the test either in headless or non-headless mode:

* `UPDATE_WC=1 npx wc-e2e test:e2e ./tests/e2e/specs/smoke-tests/update-woocommerce.test.js`
* `UPDATE_WC=1 npx wc-e2e test:e2e-dev ./tests/e2e/specs/smoke-tests/update-woocommerce.test.js`

Then:

* Verify the tests pass
* Login to the test environment and navigate to `Plugins > Installed Plugins` page
* Verify the plugin you defined in the setup above has been installed and activated as expected

### CODE REVIEW

File: tests/e2e/env/utils/get-plugin-zip.js [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+/**
+ * Download the zip file from a remote location.
+ *
+ * @param {string} fileUrl The URL where the zip file is located.
+ * @param {string} downloadPath The location where to download the zip to.
+ * @returns {Promise<void>}
+ */
+const downloadZip = async ( fileUrl, downloadPath ) => {
```
Comment (rrennick) [inline tests/e2e/env/utils/get-plugin-zip.js:61]: Also useful in to other projects and could be exported.

**Review by rrennick — CHANGES_REQUESTED** [review]
@zhongruige This looks fantastic. It worked great locally. I tested with Action Scheduler 3.2.0 and 3.2.1. I found a few nitpick items that should be fixed but otherwise this is ready to go.

File: tests/e2e/env/utils/get-plugin-zip.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+};
+
+/**
+ * Check the zip file for any nested zips. If one is found, extract it.
+ *
+ * @param {string} zipFilePath The location of the zip file.
+ * @param {string} savePath The location where to save a nested zip if found.
+ * @returns {string} The path where the zip file is located.
+ */
+const checkZip = async ( zipFilePath, savePath ) => {
```
[the remark this answers is not in the document]
Comment (zhongruige) [reply tests/e2e/env/utils/get-plugin-zip.js:39]: Oh I much prefer that name! I will make that change.

File: tests/e2e/env/utils/get-plugin-zip.js (hunk 1 above)
Comment (zhongruige) [reply tests/e2e/env/utils/get-plugin-zip.js:61]: Good call, I will export this as well as `checkNestedZip` 👍

File: .github/workflows/smoke-test-daily.yml [hunk 3]
```diff
@@ -38,5 +38,7 @@ jobs:
           E2E_RETEST: 1
           E2E_SLACK_TOKEN: ${{ secrets.SMOKE_TEST_SLACK_TOKEN }}
           E2E_SLACK_CHANNEL: ${{ secrets.SMOKE_TEST_SLACK_CHANNEL }}
+          UPDATE_WC: 1
         run: |
+          npx wc-e2e test:e2e ./tests/e2e/specs/smoke-tests/update-woocommerce.test.js
```
[the remark this answers is not in the document]
Comment (zhongruige) [reply .github/workflows/smoke-test-daily.yml:43]: Ah, yes, good call out. Let me make that change.

File: tests/e2e/env/CHANGELOG.md [hunk 4]
```diff
@@ -1,6 +1,8 @@
 # Unreleased
 
 - `updateReadyPageStatus` utility to update the status of the ready page
+- Added plugin upload functionality util that provides a method to pull a plugin zip from a remote location
+  - `getRemotePluginZip( fileUrl )` to get the remote zip. Returns the filepath of the zip location.
```
[the remark this answers is not in the document]
Comment (zhongruige) [reply tests/e2e/env/CHANGELOG.md:5]: Good idea, I added in some instructions around using them. Let me know what you think: https://github.com/woocommerce/woocommerce/pull/30291/files#diff-fb9ea0584cd5ee98bcca789f51520e4c7950d7d6a1785d81d2d81e39d71a94aeR183

**Comment by zhongruige** [comment]
Thank you for the thorough review @rrennick! I've pushed up those changes. I'll be interested in your thoughts on the `README` content, so let me know if you feel there's anything else I should add or clarify there.

File: tests/e2e/env/CHANGELOG.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (rrennick) [reply tests/e2e/env/CHANGELOG.md:5]: > Let me know what you think:

Looks good.

**Review by rrennick — APPROVED** [review]
@zhongruige This tested great again.

**Comment by rrennick** [comment]
It looks like we have a failing E2E test. I don't think this PR is responsible but we should investigate before merging.

**Comment by zhongruige** [comment]
> It looks like we have a failing E2E test. I don't think this PR is responsible but we should investigate before merging.

Yep, it's actually been failing pretty consistently. I opened this PR to work on some changes yesterday since the test otherwise passes fine locally: https://github.com/woocommerce/woocommerce/pull/30330

**Pull request merged**